Security/Http/Authenticator/OAuthAuthenticator.php

```diff
@@ -0,0 +1,250 @@
+    private $successHandler;
+
+    /**
+     * @var AuthenticationFailureHandlerInterface|null
+     */
+    private $failureHandler;
+
+    /**
+     * @var mixed[]
+     */
+    private $rawToken;
+
+    /**
+     * @var string
+     */
+    private $resourceOwnerName;
+
+    /**
+     * @var string
+     */
+    private $refreshToken;
+
+    /**
+     * @var string
+     */
+    private $createdAt;
+
+    /**
+     * Constructor.
+     *
+     * @param HttpUtils                                  $httpUtils
+     * @param OAuthAwareUserProviderInterface            $userProvider
+     * @param ResourceOwnerMapInterface                  $resourceOwnerMap
+     * @param string[]                                   $checkPaths
+     * @param AuthenticationSuccessHandlerInterface|null $successHandler
+     * @param AuthenticationFailureHandlerInterface|null $failureHandler
+     */
+    public function __construct(
+        HttpUtils $httpUtils,
+        OAuthAwareUserProviderInterface $userProvider,
+        ResourceOwnerMapInterface $resourceOwnerMap,
+        array $checkPaths,
+        ?AuthenticationSuccessHandlerInterface $successHandler = null,
+        ?AuthenticationFailureHandlerInterface $failureHandler = null
+    ) {
+        $this->failureHandler = $failureHandler;
+        $this->successHandler = $successHandler;
+        $this->checkPaths = $checkPaths;
+        $this->resourceOwnerMap = $resourceOwnerMap;
+        $this->userProvider = $userProvider;
+        $this->httpUtils = $httpUtils;
+    }
+
+    public function supports(Request $request): bool
+    {
+        foreach ($this->checkPaths as $checkPath) {
+            if ($this->httpUtils->checkRequestPath($request, $checkPath)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Authenticate.
+     *
+     * @param Request $request
+     *
+     * @return PassportInterface
+     *
+     * @throws AuthenticationException
+     * @throws LazyResponseException
+     */
+    public function authenticate(Request $request): PassportInterface
+    {
+        [$resourceOwner, $checkPath] = $this->resourceOwnerMap->getResourceOwnerByRequest($request);
+
+        if (!$resourceOwner instanceof ResourceOwnerInterface) {
+            throw new AuthenticationException('No resource owner match the request.');
+        }
+
+        if (!$resourceOwner->handles($request)) {
+            throw new AuthenticationException('No oauth code in the request.');
+        }
+
+        // If resource owner supports only one url authentication, call redirect
+        if ($request->query->has('authenticated') && $resourceOwner->getOption('auth_with_one_url')) {
+            $request->attributes->set('service', $resourceOwner->getName());
+
+            throw new LazyResponseException(new RedirectResponse(sprintf('%s?code=%s&authenticated=true', $this->httpUtils->generateUri($request, 'hwi_oauth_connect_service'), $request->query->get('code'))));
```

Comment:
What's the purpose of the Exception? `\HWI\Bundle\OAuthBundle\Security\Http\Firewall\OAuthListener::attemptAuthentication` just returns the `RedirectResponse`.

Reply:
`AuthenticatorInterface::authenticate()` [must always return](https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Security/Http/Authenticator/AuthenticatorInterface.php#L51) `PassportInterface`, so using `LazyResponseException` is the only way to return specific response to the user.